export/mlab_export.py

```diff
@@ -0,0 +1,591 @@
+    'ignored_experiments', [], 'List of experiment names to ignore. Experiment '
+    'must be in "slice.site" form not "site_slice".')
+flags.DEFINE_bool(
+    'pretty_json', None, 'Add extra indenting to json output (for debugging).')
+flags.DEFINE_string(
+    'output_dir', EXPORT_DIR, 'Root directory of json output files.')
+flags.DEFINE_string(
+    'output', None, 'Name of json output file. Set automatically if not given.')
+flags.DEFINE_string(
+    'export_metrics', METRIC_MAP_CONF, 'File name with metric map. The metric '
+    'map defines canonical metric names for raw, metric names taken from '
+    'collectd RRD files.')
+flags.DEFINE_bool(
+    'verbose', False, 'Increase verbosity level.')
+flags.DEFINE_bool(
+    'show_nagios', False, 'Shows collectd-nagios commands to monitor metrics.')
+flags.DEFINE_bool(
+    'show_rrdfile', False, 'Shows the RRD files opened during export.')
+flags.DEFINE_bool(
+    'show_metric', False, 'Shows the canonical metric names during export.')
+flags.DEFINE_bool(
+    'show_rawmetric', False, 'Shows the raw metric name before translating to '
+    'the canonical name. This can be helpful to know before adding new metrics '
+    'to the export metrics configuration file.')
+flags.DEFINE_bool(
+    'show_skipped', False, 'Shows the raw metric names that are not exported. '
+    'This option may be helpful when adding new metrics to the export_metrics.')
+flags.DEFINE_bool(
+    'update', True, 'Update timestamps on successful export. Update is always '
+    'disabled when any --show_* option is enabled.')
+flags.DEFINE_bool(
+    'compress', False,
+    'Compresses output and adds .gz extension to output filename.')
+
+
+class Error(Exception):
+  """Base error type for this file."""
+
+
+class TimeOptionError(Error):
+  """An error related to export times or ranges."""
+
+
+def init_global():
+  global HOSTNAME
+  # NOTE: This should be the hostname of root context, not slice context.
+  HOSTNAME = socket.gethostname()
+  logging.basicConfig(format='%(message)s', level=logging.INFO)
+
+
+class LastTime(object):
+  """Manages interactions with export timestamp."""
+
+  def __init__(self, filename):
+    self._filename = filename
+    self._lock_handle = None
+    self._first_creation = not os.path.exists(self._filename)
+
+  def __enter__(self):
+    self.open()
+    return self
+
+  def __exit__(self, *unused_args):
+    self.close()
+
+  def open(self):
+    """Opens the timestamp file, and creates the file if it does not exist."""
+    if self._first_creation:
+      # Since the file does not exist, use mode 'w' to create it.
+      self._lock_handle = open(self._filename, 'w')
+    else:
+      # File *does* exist, so use mode 'r' to preserve the mtime on the file.
+      self._lock_handle = open(self._filename, 'r')
+
+  def close(self):
+    """Closes the timestamp file."""
+    if self._lock_handle is not None:
+      self._lock_handle.close()
+
+  def get_mtime(self):
```

Comment:
Great points. This was an interesting one to tease apart.

My current thinking is that I was trying to do too much with the timestamp file. The LastTime class was using it both to record the last export time *and* as a file lock for the running process. Oh, right, and why not make it a context-aware object too? :-)

I like your suggestion to make open/close/lock hidden within the context management functions. I saw two options to achieve this.

1. perform the open/lock in `__enter__`. However, it turns out that to catch exceptions that occur in `__enter__`, the entire `with` block would need to be in a `try/except`. This is two levels of indenting and it felt a little funny to wrap that much code with an except.
2. perform the open/lock in `__init__` so that the object can be created and the lock attempted before the `with` block so we can catch any exceptions before continuing. This requires taking complex actions in `__init__`, which I regard as an anti-pattern for testing. So, I didn't want to do this either.

So, in the interest of simplicity, I removed the get_mtime & update_mtime methods and made them plain functions. Then I created a basic LockFile class to manage the advisory lock. I think the separation is positive, but there may yet be improvements to the new parts.

How does it look now? (Total change across two commits): https://github.com/stephen-soltesz/collectd-mlab/compare/99b050b86b4fdbcd816db86a265165755ac400cd...6cda59abd3c8df863f285698acdae5d34f5e92f8